### PR TITLE
Ame 19

### PR DESCRIPTION
**Pull request by MHekert**

### CODE REVIEW

File: src/controllers/mode/addModeController.ts [hunk 1]
```diff
@@ -18,7 +17,7 @@ router.put(
 			const savedMode = mode.save();
 			res.status(200).send({ mode: await savedMode });
 		} catch (err) {
-			putErrorHandler(err, res);
+			errorHandler(err, res);
 		}
 	}
 );
```
Comment (JBudny) [inline src/controllers/mode/addModeController.ts:25]: As long as you don't use `next` parameter, you can omit it in the function because it's an optional parameter.

File: src/controllers/mode/addModeController.ts [hunk 2]
```diff
@@ -3,8 +3,7 @@ import { validationResult, body } from 'express-validator/check';
 import Mode from '../../models/mode/mode';
 import Revision from '../../models/mode/revision';
 import modelFromRequest from '../../util/modelFromRequest';
-import putErrorHandler from '../helpers/putErrorHandler';
-
+import errorHandler from '../helpers/errorHandler';
```
Comment (JBudny) [inline src/controllers/mode/addModeController.ts:6]: One empty line after imports and then after const would look pleasing

File: src/controllers/ModeAction/downvote.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { Router, Request, Response } from 'express';
+import errorHandler from '../helpers/errorHandler';
+import { setDownvote } from '../../models/modeAction';
+import { decPoints } from '../../models/mode/mode';
```
Comment (JBudny) [inline src/controllers/ModeAction/downvote.ts:4]: According to `import/newline-after-import` eslint rule, one empty line after all imports would look pleasing

File: src/controllers/ModeAction/helpers/validateRequest.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { Router, Request, Response, NextFunction } from 'express';
+import { validationResult, body } from 'express-validator/check';
+import errorHandler from '../../helpers/errorHandler';
+import isUserLoggedIn from '../../middleware/isUserLoggedIn';
```
Comment (JBudny) [inline src/controllers/ModeAction/helpers/validateRequest.ts:4]: One empty line after imports would look pleasing

File: src/controllers/ModeAction/setFavorite.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import { Router, Request, Response } from 'express';
+import errorHandler from '../helpers/errorHandler';
+import { setFavorite } from '../../models/modeAction';
+import { incFavorite } from '../../models/mode/mode';
```
Comment (JBudny) [inline src/controllers/ModeAction/setFavorite.ts:4]: One empty line after imports would look pleasing

File: src/controllers/ModeAction/unsetFavorite.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import { Router, Request, Response } from 'express';
+import errorHandler from '../helpers/errorHandler';
+import { unsetFavorite } from '../../models/modeAction';
+import { decFavorite } from '../../models/mode/mode';
```
Comment (JBudny) [inline src/controllers/ModeAction/unsetFavorite.ts:4]: One empty line after imports would look pleasing

File: src/controllers/ModeAction/unsetVote.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import { Router, Request, Response } from 'express';
+import errorHandler from '../helpers/errorHandler';
+import { unsetVote } from '../../models/modeAction';
+import { incPoints, decPoints } from '../../models/mode/mode';
```
Comment (JBudny) [inline src/controllers/ModeAction/unsetVote.ts:4]: One empty line after imports would look pleasing

File: src/controllers/ModeAction/upvote.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import { Router, Request, Response } from 'express';
+import errorHandler from '../helpers/errorHandler';
+import { setUpvote } from '../../models/modeAction';
+import { incPoints } from '../../models/mode/mode';
```
Comment (JBudny) [inline src/controllers/ModeAction/upvote.ts:4]: One empty line after imports would look pleasing

File: src/controllers/revision/addRevisionController.ts [hunk 9]
```diff
@@ -17,7 +17,7 @@ router.put(
 			await revision.insertToMode(req.body.modeId);
 			res.status(200).send(revision);
 		} catch (err) {
-			putErrorHandler(err, res);
+			errorHandler(err, res);
 		}
 	}
 );
```
Comment (JBudny) [inline src/controllers/revision/addRevisionController.ts:23]: As long as you don't use `next `parameter, you can omit it in the function because it's an optional parameter.

File: tests/controllers/modeAction/setFavoriteTest.ts [hunk 10]
```diff
@@ -0,0 +1,63 @@
+process.env.NODE_ENV = 'test';
+import { expect, request, use } from 'chai';
+import { connection, Types } from 'mongoose';
```
Comment (JBudny) [inline tests/controllers/modeAction/setFavoriteTest.ts]: 'Types' is declared but its value is never read.

File: tests/controllers/modeAction/setFavoriteTest.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+		const updatedMode = await Mode.findById(mode._id);
+		expect(updatedMode.favorites).to.be.equal(mode.favorites + 1);
+	});
+
+	it('should return status code 304 and decrease points by 1', async () => {
+		await request(app)
+			.post('/mode/action/setfavorite')
+			.set('content-type', 'application/json')
+			.send({ modeId: mode._id });
+		const res = await request(app)
```
Comment (JBudny) [inline tests/controllers/modeAction/setFavoriteTest.ts:55]: Is that request executed twice on purpose?

File: tests/controllers/modeAction/downvoteTest.ts [hunk 12]
```diff
@@ -0,0 +1,69 @@
+process.env.NODE_ENV = 'test';
+import { expect, request, use } from 'chai';
+import { connection, Types } from 'mongoose';
```
Comment (JBudny) [inline tests/controllers/modeAction/downvoteTest.ts]: 'Types' is declared but its value is never read.

File: tests/controllers/modeAction/unsetFavoriteTest.ts [hunk 13]
```diff
@@ -0,0 +1,61 @@
+process.env.NODE_ENV = 'test';
+import { expect, request, use } from 'chai';
+import { connection, Types } from 'mongoose';
```
Comment (JBudny) [inline tests/controllers/modeAction/unsetFavoriteTest.ts]: 'Types' is declared but its value is never read.

File: tests/controllers/modeAction/unsetVoteTest.ts [hunk 14]
```diff
@@ -0,0 +1,61 @@
+process.env.NODE_ENV = 'test';
+import { expect, request, use } from 'chai';
+import { connection, Types } from 'mongoose';
```
Comment (JBudny) [inline tests/controllers/modeAction/unsetVoteTest.ts]: 'Types' is declared but its value is never read.

File: tests/controllers/modeAction/upvoteTest.ts [hunk 15]
```diff
@@ -0,0 +1,68 @@
+process.env.NODE_ENV = 'test';
+import { expect, request, use } from 'chai';
+import { connection, Types } from 'mongoose';
```
Comment (JBudny) [inline tests/controllers/modeAction/upvoteTest.ts]: 'Types' is declared but its value is never read.

File: tests/models/modeActionTest.ts [hunk 16]
```diff
@@ -1,74 +1,81 @@
 import { describe, it } from 'mocha';
 import { expect } from 'chai';
 import { connection } from 'mongoose';
-import ModeAction from '../../src/models/modeAction';
+import ModeAction, { setDownvote, setUpvote, setFavorite, unsetFavorite, unsetVote } from '../../src/models/modeAction';
 import { MONGODB_URI_TEST } from '../../src/util/secrets';
-import { dummyModeAction, getDummyModeAction } from '../dummyData/dummyModeAction';
+import { dummyModeAction, dummyModeActionBody, getDummyModeAction } from '../dummyData/dummyModeAction';
```
Comment (JBudny) [inline tests/models/modeActionTest.ts:6]: One empty line after imports would look pleasing

File: tests/models/mode/modeTest.ts [hunk 17]
```diff
@@ -1,7 +1,8 @@
 import { describe, it } from 'mocha';
 import { expect } from 'chai';
 import { connection } from 'mongoose';
-import Mode from '../../../src/models/mode/mode';
+import Mode, { incPoints, decPoints, incFavorite } from '../../../src/models/mode/mode';
+
```
Comment (JBudny) [inline tests/models/mode/modeTest.ts]: Shouldn't that empty line be after all imports?

File: src/controllers/ModeAction/downvote.ts [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+const router: Router = Router();
+
+router.post('/', async (req: Request, res: Response) => {
+	try {
+		const modeAction = await setDownvote(req.user, req.body.modeId);
+		if (!modeAction || modeAction.upvote === undefined) {
```
Comment (JBudny) [inline src/controllers/ModeAction/downvote.ts]: INFO:
There could be the case that this condition could behave a little unpredictable.
According to this eslint rule [no-undefined](https://eslint.org/docs/rules/no-undefined), provided that you use ES3 or older version of Javascript specification, you should compare variables to `undefined` like below:
`if (!modeAction || typeof modeAction.upvote === "undefined") {...}`

Back then the popular solution of that problem was to obtain undefined by using the `void` operator like this:

`if (name === void 0) {...}`

Unfortunately, this operator wasn't self-explanatory and from that reason, it was advised to wrap it inside a function called eg.: `isUndefined` and many third-party libraries provided that.

However, the problem of re-assignable `undefined` value was fixed back in 2009 and then the popular libraries were changed from obtaining `undefined` value from `void 0` to just compare passed argument with `undefined` directly.
Example: [Lodash commit form 2012](https://github.com/lodash/lodash/blob/9fc5044cf985f1d8f020c20f4e96541cef431825/lodash.js#L2063)

In conclusion, as long as you don't use ES3 or older version of Javascript specification you can compare variables directly to `undefined` as you did above. If you'd like to use ES3 or older language specification you should use the method advised by eslint or find a library that uses that method.

File: tests/controllers/modeAction/setFavoriteTest.ts (hunk 11 above)
Comment (MHekert) [reply tests/controllers/modeAction/setFavoriteTest.ts:55]: Yes, it tests that specific scenario.

**Review by JBudny — APPROVED** [review]
